pkg/lease/manager.go

```diff
@@ -0,0 +1,228 @@
+		lease.ObjectMeta.OwnerReferences = []metav1.OwnerReference{*owner}
+		lease.Spec.HolderIdentity = &l.holderIdentity
+		lease.Spec.LeaseDurationSeconds = pointer.Int32(int32(leaseDuration.Seconds()))
+		lease.Spec.RenewTime = &currentTime
+		if err := l.Client.Update(ctx, lease); err != nil {
+			log.Errorf("Failed to update the lease. obj %s error: %v", obj.GetName(), err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (l *manager) invalidateLease(ctx context.Context, obj client.Object) error {
+	log.Info("invalidating lease")
+	lease, err := l.getLease(ctx, obj)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+	if err := l.Client.Delete(ctx, lease); err != nil {
+		log.Error(err, "failed to delete lease to be invalidated")
+		return err
+	}
+
+	return nil
+}
+
+func makeExpectedOwnerOfLease(obj client.Object) *metav1.OwnerReference {
+
+	return &metav1.OwnerReference{
+		APIVersion: obj.GetObjectKind().GroupVersionKind().Version,
+		Kind:       obj.GetObjectKind().GroupVersionKind().Kind,
+		Name:       obj.GetName(),
+		UID:        obj.GetUID(),
+	}
+}
+
+func leaseDueTime(lease *coordv1.Lease) time.Time {
+	return lease.Spec.RenewTime.Time.Add(time.Duration(*lease.Spec.LeaseDurationSeconds) * time.Second)
+}
+
+func needUpdateOwnedLease(lease *coordv1.Lease, currentTime metav1.MicroTime, requestedLeaseDuration time.Duration) (bool, bool) {
+
+	if lease.Spec.RenewTime == nil || lease.Spec.LeaseDurationSeconds == nil {
+		log.Info("empty renew time or duration in sec")
+		return true, true
+	}
+	dueTime := leaseDueTime(lease)
+
+	// if lease expired right now, then both update the lease and the acquire time (second rvalue)
+	// if the acquire time has been previously nil
+	if dueTime.Before(currentTime.Time) {
+		return true, lease.Spec.AcquireTime == nil
+	}
+
+	deadline := currentTime.Add(requestedLeaseDuration)
+
+	// about to expire, update the lease but not the acquired time (second value)
+	return dueTime.Before(deadline), false
+}
+
+func isValidLease(lease *coordv1.Lease, currentTime time.Time) bool {
+
+	if lease.Spec.RenewTime == nil || lease.Spec.LeaseDurationSeconds == nil {
+		return false
+	}
+
+	renewTime := (*lease.Spec.RenewTime).Time
+	dueTime := leaseDueTime(lease)
+
+	// valid lease if: due time not in the past and renew time not in the future
+	return !dueTime.Before(currentTime) && !renewTime.After(currentTime)
+}
+
+func (l *manager) getLease(ctx context.Context, obj client.Object) (*coordv1.Lease, error) {
+	log.Info("getting lease")
+	nName := apitypes.NamespacedName{Namespace: l.namespace, Name: generateLeaseName(obj)}
+	lease := &coordv1.Lease{}
+
+	if err := l.Client.Get(ctx, nName, lease); err != nil {
+		if !errors.IsNotFound(err) {
+			l.log.Error(err, "couldn't fetch lease")
+		}
+		return nil, err
+	}
+
+	return lease, nil
+}
+
+func generateLeaseName(obj client.Object) string {
+	return fmt.Sprintf("%s_%s", obj.GetObjectKind().GroupVersionKind().Kind, obj.GetName())
```

Comment:
underscores are not allowed for names, please use minus sign: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names